src/tactic/linarith.lean

```diff
@@ -98,6 +98,103 @@ by rw [←h3, mul_assoc, mul_div_comm, h2, ←mul_assoc, h1, mul_comm, one_mul]
 
 end lemmas
 
+/--
+A linear expression is a list of pairs of variable indices and coefficients.
+
+Some functions on `linexp` assume that `n : ℕ` occurs at most once as the first element of a pair,
+and that the list is sorted in decreasing order of the first argument.
+This is not enforced by the type but the operations here preserve it.
+-/
+@[reducible]
+def linexp : Type := list (ℕ × ℤ)
+end linarith
+
+/--
+A map `ℕ → ℤ` is converted to `list (ℕ × ℤ)` in the obvious way.
+This list is sorted in decreasing order of the first argument.
+-/
+meta def native.rb_map.to_linexp (m : rb_map ℕ ℤ) : linarith.linexp :=
+m.to_list
+
+namespace linarith
+namespace linexp
+
+/--
+Add two `linexp`s together componentwise.
+Preserves sorting and uniqueness of the first argument.
+-/
+def add : linexp → linexp → linexp
+| [] a := a
+| a [] := a
+| (a@(n1,z1)::t1) (b@(n2,z2)::t2) :=
+  match cmp n1 n2 with
+  | ordering.lt := b::add (a::t1) t2
+  | ordering.gt := a::add t1 (b::t2)
+  | ordering.eq := let sum := z1 + z2 in if sum = 0 then add t1 t2 else (n1, sum)::add t1 t2
+  end
+
+/-- `l.scale c` scales the values in `l` by `c` without modifying the order or keys. -/
+def scale (c : ℤ) (l : linexp) : linexp :=
+if c = 0 then []
+else l.map $ λ ⟨n, z⟩, (n, z*c)
+
+/--
+`l.get n` returns the value in `l` associated with key `n`, if it exists, and `none` otherwise.
+This function assumes that `l` is sorted in decreasing order of the first argument,
+that is, it will return `none` as soon as it finds a key smaller than `n`.
+-/
+def get (n : ℕ) : linexp → option ℤ
+| [] := none
+| ((a, b)::t) := match cmp a n with
+  | ordering.lt := none
+  | ordering.gt := get t
+  | ordering.eq := some b
+  end
+
+/--
+`l.contains n` is true iff `n` is the first element of a pair in `l`.
+-/
+def contains (n : ℕ) : linexp → bool := option.is_some ∘ get n
+
+/--
+`l.zfind n` returns the value associated with key `n` if there is one, and 0 otherwise.
+-/
+def zfind (n : ℕ) (l : linexp) : ℤ :=
+match l.get n with
+| none := 0
+| some v := v
+end
+
+/--
+Defines a lex ordering on `linexp`. This function is performance critical.
+-/
+def lt : linexp → linexp → bool
+| [] [] := ff
+| [] _ := tt
+| _ [] := ff
+| ((n1,z1)::t1) ((n2,z2)::t2) := match cmp n1 n2 with
+  | ordering.lt := tt
+  | ordering.gt := ff
+  | ordering.eq := match cmp z1 z2 with
+    | ordering.lt := tt
+    | ordering.gt := ff
+    | ordering.eq := lt t1 t2
+    end
+  end
+
+end linexp
+
+-- these instances are local to this file so that they don't accidentally leak to other uses of `list (ℕ × ℤ)`.
+
+/-- A local `has_lt` instance, implemented by `linarith.linexp.lt`. -/
+local attribute [instance]
+def linexp.has_lt : has_lt linexp := ⟨↑linexp.lt⟩
+
+/-- A local decidability instance. -/
+local attribute [instance]
+def linexp.lt_decidable : decidable_rel ((<) : linexp → linexp → Prop) :=
+λ a b, decidable_of_bool (linexp.lt a b) $ iff.refl _
```

Comment:
```suggestion
local attribute [instance, reducible]
def linexp.has_lt : has_lt linexp := ⟨λ a b, a.lt b⟩
```

Reply:
You might also just write `.lt` instead of `<` the one time it is used.

Reply:
Indeed, you're right. I was expecting to have `rb_set linexp` which would need the instance, but the `linexp`s get bundled in another structure.